Add apt to dpup, as a layer above the main SFS

dimkr opened the pull request:
The wonders a small 19 MB SFS can do 😺 

![update](https://user-images.githubusercontent.com/1471149/149672037-190a0448-ea05-48f9-8f5e-b2f5c2462747.png)

![vlc](https://user-images.githubusercontent.com/1471149/149672025-37e0d885-28b7-4fc1-b771-5e2930344ab3.png)

![synaptic](https://user-images.githubusercontent.com/1471149/149672028-5a7a43a2-c313-4bfe-883e-05dd893e1fe2.png)

![abiword](https://user-images.githubusercontent.com/1471149/149672019-d9219440-51ae-4477-b5f9-2b3727270e7d.png)

peabee reviewed (commented):
I guess I'm surprised at how complicated and pervasive these apt changes need to be...... 
neither pkg nor xbps (as in VoidPup) needed such extensive Woof-CE changes.....
On a philosophical point I think that initiatives like this that only affect a single build should use other Woof-CE "mechanisms" such as a specific specialized .pet to introduce just this feature and/or use of a pinstall in such a .pet or code in _00build.conf EXTRA_COMMANDS to achieve what is needed to incorporate the feature.

dimkr commented:
> neither pkg nor xbps (as in VoidPup) needed such extensive Woof-CE changes.....

The changes are not that big - most of the diff is the addition of users and groups. woof-code/support/bdrv.sh, the only new file, is pretty short, and I think it's really clear.

The customization of the Debian installation is needed to reduce its size, and prevent conflicts with Puppy. For example, the user must not `apt upgrade coreutils`, because that can easily replace /bin/cp, a symlink to /bin/busybox, which will replace busybox with cp, and the system will no longer boot.

VoidPup and any other Puppy that uses a foreign package manager will need the same kind of protection.

After this explanation, do you still find this PR "complicated and pervasive" (quote)?

EDIT: forgot to address this:

>  I think that initiatives like this that only affect a single build should use other Woof-CE "mechanisms" such as a specific specialized .pet

It's impossible. `apt` and its dependencies must sit below the main SFS, and be separate. And this "bdrv" must be built after the contents of rootfs-complete are known. Existing mechanisms, like pinstall.sh, package templates, ydrv, EXTRA_COMMANDS, you name it - won't work here.

mavrothal commented:
Interesting!
Do you know if changes in /etc/{group,passwd,shadow} affect spot user and the run-as-spot utility?
Also, there is no change in init to load bdrv. Is that intentional?

01micko commented on woof-code/3builddistro:
```diff
@@ -962,20 +963,28 @@ fi
 if [ "$BUILD_SFS" = 'yes' ]; then
 	sh $MWD/support/files2delete.sh rootfs-complete
+
+	# frugalify puts bdrv below the main SFS
```
@mavrothal

01micko replied:
Doesn't touch `init`

dimkr replied:
Testing apt when booting with initrd.gz - so far, all basic tests passed. I'll add initrd.gz support to this PR soon.

dimkr commented:
> Interesting! Do you know if changes in /etc/{group,passwd,shadow} affect spot user and the run-as-spot utility?

No, spot works. (dpup has many applications running as spot and they all work; Firefox, PulseAudio and Blueman, among others, run as spot. But this is unsurprising, because the things that make spot work are all in the main SFS and always override the Debian configuration files).

I added all users and groups defined in a clean Debian `debootstrap`. I fixed every collision between Puppy and Debian by giving the Puppy user/group a different ID. For example, `sshd` had UID 33: I changed it to 1003 so the `www-data` user can have the same ID it has on Debian.

Packages with files that must be owned by a specific user or group can break, but only if these packages contain executables that run as non-root users (for example, the `disk` group is no longer 100).

(apt needs the `_apt` user, but I decided to sync all users and groups in the name of compatibility. It's safe to add more users and groups, but a missing user can break an application.)

> Also, there is no change in init to load bdrv. Is that intentional?

Yes - for now, this feature only works with https://github.com/puppylinux-woof-CE/frugalify. https://github.com/puppylinux-woof-CE/frugalify/commit/62e6efaabcb42ffd8dbd79af1aab2d7113d14088 changes the SFS sorting algorithm, so bdrv is below the main SFS, then everything else.

For this "bdrv" concept to work with initrd.gz too, `init` must place bdrv below the main SFS.

dimkr committed:
add apt to dpup, as a layer below the main SFS

dimkr commented:
> For this "bdrv" concept to work with initrd.gz too, init must place bdrv below the main SFS.

Done! Please review.

I had to move some apt dependencies from devx to the main SFS, because the ISO images don't come with devx.

~~I'll publish a test ISO soon,~~ so you can try it out. https://github.com/dimkr/woof-CE/releases/download/dpup-8.0-apt6/dpup-8.0.iso

dimkr committed:
update dpup release notes

dimkr committed:
make it possible to build dpup without bdrv

dimkr committed:
optionally symlink directories at / to /usr

dimkr commented:
https://github.com/dimkr/woof-CE/releases/download/vanilladpup-9.9.9-apt/vanilladpup-9.9.9.iso

Together with #2771:

![wine32](https://user-images.githubusercontent.com/1471149/150183131-4a0aef8f-c52f-40db-bfbf-3d9268397c25.png)

![links](https://user-images.githubusercontent.com/1471149/150183143-f3596626-60cc-47f2-96aa-b1a100c534c6.png)

![steam](https://user-images.githubusercontent.com/1471149/150183823-1c6082d8-45cd-478d-9013-f3bb0c94caa0.png)

As you can see, it's possible to install 32-bit libraries and applications that rely on them, like Wine and Steam.

dimkr committed:
Merge branch 'feature/usr-merge-v2' of https://github.com/dimkr/woof-CE into feature/dpup-apt

dimkr committed:
fix missing source

dimkr committed:
move packages needed to install Wine out of devx

dimkr committed:
Merge branch 'testing' of https://github.com/puppylinux-woof-CE/woof-CE into feature/dpup-apt

dimkr commented on initrd-progs/0initrd/init:
```diff
@@ -76,13 +76,15 @@ export KERNELVER="$(uname -r)"
 export DISTRODESC="${DISTRO_NAME} ${DISTRO_VERSION} - Linux ${KERNELVER} - `uname -m`"
 
 #precaution - if DISTRO_SPECS was not processed by 3builddistro...
+[ ! "$DISTRO_BDRVSFS" ] && DISTRO_BDRVSFS="bdrv_${DISTRO_FILE_PREFIX}_${DISTRO_VERSION}.sfs"
```
@mavrothal Did I mess up initrd in any way? This code is messy and I want to make sure I didn't break anything.

dimkr committed:
restore missing icons in Synaptic

dimkr committed:
add support for automatic menu refresh when a .desktop is added or deleted

dimkr committed:
Merge branch 'feature/fixmenusd' into feature/dpup-apt

dimkr committed:
fix exit of fixmenusd after one apt install

dimkr commented:
https://github.com/dimkr/woof-CE/releases/download/vanilladpup-9.9.9-apt4/vanilladpup-9.9.9.iso

Missing icons are back:

![icons](https://user-images.githubusercontent.com/1471149/150502930-7a1ec934-4d8c-4c2e-a6bf-94876c390a08.png)

Installation of Xfce after first boot of the ISO, without devx, works now:

![start](https://user-images.githubusercontent.com/1471149/150502984-44bb9000-574b-43fa-aa51-e4ec2d6955ff.png)

Xfce even works **(and works under Xwayland!)**:

![xfce](https://user-images.githubusercontent.com/1471149/150502937-e8e6c7d4-d699-4bf1-a91a-2c1dc819fe60.png)

After setting the GTK+ theme, icon theme and wallpaper it looks pretty usable:
 
![xfce2](https://user-images.githubusercontent.com/1471149/150503551-a3f736bf-9fb0-413b-a515-91af5ddeb700.png)

Same with KDE:

![kde](https://user-images.githubusercontent.com/1471149/150513655-c869ca99-cd64-41e9-b628-6cefeda28de1.png)

And, the menu JWM is updated automatically and silently after every `apt install`.

IMHO this PR is useful and complete enough to merge, and everything is optional and disabled by default.

dimkr committed:
cleanup

dimkr committed:
Merge branch 'testing' of https://github.com/puppylinux-woof-CE/woof-CE into feature/dpup-apt

dimkr committed:
move ucf out of dpup devx

dimkr committed:
fix typo

dimkr committed:
make bdrv smaller

dimkr committed:
add passwd to fix apt failure when trying to add users or groups

dimkr committed:
in ISO builds, handle bdrv like docx and nlsx

dimkr committed:
use the same /etc/{group,passwd,shadow} as before for all builds except dpup

dimkr committed:
document the purpose of bdrv

dimkr committed:
bump the version number of dpup bullseye

dimkr commented:
With 8a2d829 and 6c5f14e, all builds except dpup are unaffected by this PR.

72cc171 makes it safe to load bdrv above the main SFS, so it's unnecessary to change initrd to load bdrv first.

All basic tests passed (VLC, GIMP, Xfce, KDE, gdm3, and now also LibreOffice):

![writer](https://user-images.githubusercontent.com/1471149/150635328-49cb8432-0bf5-4d3c-bfb6-5e4231a52973.png)

@peabee @mavrothal @01micko Objections?
